### ISSUE TITLE
complex64 multipled by float32 produces a complex128

### ISSUE THREAD

**Issue by sjperkins** [pydata/numexpr#155]
This occurs on numexpr 2.4, Ubuntu 14.04

The following evaluation results in a complex128 being produced, rather than a complex64. I don't think this follows the casting rules, but I may be interpreting them incorrectly.

``` python
import numexpr as ne
import numpy as np

A = (np.random.random(20) + np.random.random(20)*1j).astype(np.complex64)
B = np.random.random(20)
ne.evaluate('a*b', {'a':A,'b':B},out=A)
```

Without the **out** parameter, we could use `astype` as a stopgap. However, with an **out** parameter we can't output the result to the A array The following exception is thrown:

``` python
TypeError: Iterator requested dtype could not be cast from dtype('complex128') to dtype('complex64'), the operand 0 dtype, according to the rule 'safe'
```

**Comment by pakodekker** [comment]
@sjperkins did anything happen on this front? I see that you a pull request. Support for complex64 seems critical to me!

**Comment by sjperkins** [comment]
@pakodekker I started on it -- see the [complex64](https://github.com/sjperkins/numexpr/tree/complex64) branch on my fork. This became less critical for me and my time got taken up by other things, so I'm not actively pursuing it at present.

**Comment by robbmcleod** [comment]
I was able to get this working, you can see it on my page:

https://github.com/robbmcleod/numexpr

I will need to make some reverts to make it more compatible with the commits that have been made since Nov 2014 before it can be merged with the master branch.  I've tested it with both the numexpr/tests/testexpr.py and bench/timing_complex.py.

Perhaps more importantly I tested it with my own code that also uses pyFFTW, and it plays nice with it.  pyFFTW is really picky with regards to array alignment and dtype so I'm confident it's working.  There are some casting issues, with regards to how does double cast to complex64.

**Comment by sjperkins** [comment]
@robbmcleod Great! I'm glad the initial work didn't go to waste.

**Comment by FrancescAlted** [comment]
@robbmcleod would you please send a PR of your work so that we can use that instead of #156?  Thanks!

**Comment by robbmcleod** [comment]
There's still one outstanding problem of how to deal with real, imag, and
complex64 casts.  The AST interpreter is really bad at figuring out return
signatures, so it casts them all to double.

I'm hopeful that it can be dealt with quickly, or at least that I can
figure out a work-around/hack.


On Tue, Jul 28, 2015 at 5:51 PM, FrancescAlted <notifications@github.com>
wrote:

> @robbmcleod <https://github.com/robbmcleod> would you please send a PR of
> your work so that we can use that instead of #156
> <https://github.com/pydata/numexpr/pull/156>? Thanks!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pydata/numexpr/issues/155#issuecomment-125661134>.
>



-- 
Robert McLeod, Ph.D.
Center for Cellular Imaging and Nano Analytics (C-CINA)
Biozentrum der Universität Basel
Mattenstrasse 26, 4058 Basel
Work: +41.061.387.3225
robert.mcleod@unibas.ch
robert.mcleod@bsse.ethz.ch <robert.mcleod@ethz.ch>
robbmcleod@gmail.com

**Comment by gdementen** [comment]
At a glance, this PR reverts a few recent commits. There is probably a bad merge somewhere in there.

**Comment by robbmcleod** [comment]
Closing as this was the case for the NumExpr-3 branch.  Complex64 works as expected in NE3.

**Comment by mutricyl** [comment]
Working on https://github.com/pandas-dev/pandas/issues/21374 I am Coming across unexpected complex128 upcasting.

The above example fails with numexpr 2.10. I am not sure to understand if it is solve in numexpr 3 branch which is not release yet or if it was solved in 2.3 and some kind of regression append since. @robbmcleod should this issue be re-opened ?
